### PR TITLE
feat(portfolio-contract): claim rewards on Aave, Compound

### LINKED ISSUE #11643: [ymax] claim rewards on Aave and Compound
Add the ability to claim rewards from an Aave/Compound position in the portfolio contract
The user should be able to pass a flag while withdrawing that tells the remote account to claim rewards before withdrawing.

### PR DESCRIPTION
**Pull request by frazarshad**
closes: #11643 

## Description

Adds the following:
- functions for claiming rewards in aave and compound
- Logic for triggering these functions through the `rebalance` function
- Tests for this functionality

Adding any new addresses required has been left for a later PR

### Security / Scaling Considerations

nothing new

### Documentation Considerations

Clients should look at the test for how to use the `claim: true` option in offerArgs.

### Testing Considerations

contract test for the new user story
plus one showing USDN doesn't support it yet

### Upgrade Considerations

no changes to durable state; only behavior. and backward-compatible.
so: unremarkable

### CODE REVIEW

File: packages/portfolio-contract/src/pos-gmp.flows.ts [hunk 1]
```diff
@@ -176,7 +176,14 @@ export type EVMContext<CN extends string> = {
   lca: LocalAccount;
   gmpFee: DenomAmount;
   gmpChain: Chain<{ chainId: string }>;
-  addresses: Record<CN | 'usdc', EVMT['address']>;
+  addresses: Record<
+    | CN
+    | 'usdc'
+    | 'aaveRewardsController'
+    | 'aaveUSDC'
+    | 'compoundRewardsController',
```
Comment (dckc) [inline packages/portfolio-contract/src/pos-gmp.flows.ts]: Only `usdc` is common to all `EVMContext`s. These belong on the individual `ProtocolDetail`s.
```suggestion
```

File: packages/portfolio-contract/src/pos-gmp.flows.ts [hunk 2]
```diff
@@ -266,6 +306,21 @@ export const CompoundProtocol = {
...
+      a.compoundRewardsController,
+      CompoundRewardsController,
+    );
+    compoundRewardsController.claim(a.compound, src.remoteAddress, true);
+    const calls = session.finish();
+
+    const axelarId = ctx.axelarIds[src.chainName];
+    const target = { axelarId, remoteAddress: src.remoteAddress };
+    await sendGMPContractCall(target, calls, gmpFee, lca, gmpChain);
+  },
```
Comment (dckc) [inline packages/portfolio-contract/src/pos-gmp.flows.ts]: a few lines down where we have `EVMContext<'compound'>`, that's where the new contracts belong: `EVMContext<'compound' | 'compoundRewardsController'>`

There's evidently not enough docs to make this clear; here's hoping we improve on that.

File: packages/portfolio-contract/src/pos-gmp.flows.ts [hunk 3]
```diff
@@ -221,6 +236,21 @@ export const AaveProtocol = {
...
+      a.aaveRewardsController,
+      AaveRewardsController,
+    );
+    aaveRewardsController.claimAllRewardsToSelf([a.aaveUSDC]);
+    const calls = session.finish();
+
+    const axelarId = ctx.axelarIds[src.chainName];
+    const target = { axelarId, remoteAddress: src.remoteAddress };
+    await sendGMPContractCall(target, calls, gmpFee, lca, gmpChain);
+  },
```
Comment (dckc) [inline packages/portfolio-contract/src/pos-gmp.flows.ts]: Likewise, change `EVMContext<'aavePool'>` to `EVMContext<'aavePool' | 'aaveRewardsController' | 'aaveUSDC'>

File: packages/portfolio-contract/src/evm-facade.ts [hunk 4]
```diff
@@ -10,6 +10,8 @@ export type EVMT = {
   uint256: NatValue;
   address: `0x${string}`;
   bytes32: `0x${string}`;
+  'address[]': Array<`0x${string}`>;
+  bool: boolean;
```
Comment (dckc) [inline packages/portfolio-contract/src/evm-facade.ts:14]: aside: I looked into `functionSignature` and realized that `evm-facade.ts` is largely re-inventing parts of viem: `evm-facade.ts` has code to assemble the parts of a `functionSignature` into a string, only to have it parsed back apart in `constructContractCall` before we hand it to `viem`'s `encodeFunctionData`.

Here's hoping we can cut out that formatting and parsing.

File: packages/portfolio-contract/src/portfolio.contract.ts [hunk 5]
```diff
@@ -85,6 +85,9 @@ export type EVMContractAddresses = {
   factory: `0x${string}`;
   usdc: `0x${string}`;
   tokenMessenger: `0x${string}`;
+  aaveUSDC: `0x${string}`;
+  aaveRewardsController: `0x${string}`;
+  compoundRewardsController: `0x${string}`;
```
Comment (dckc) [inline packages/portfolio-contract/src/portfolio.contract.ts:90]: aside: This `EVMContractAddresses` type is awkward: its shape suggests the chain knows all the relevant contracts. But the way the information flows is: aave tells us the addresses on the chains where it's deployed and likewise compound etc.

File: packages/portfolio-contract/src/portfolio.flows.ts [hunk 6]
```diff
@@ -148,6 +148,7 @@ export type ProtocolDetail<
     amount: NatAmount,
     dest: AccountInfoFor[C],
   ) => Promise<void>;
+  claimRewards?: (ctx: CTX, src: AccountInfoFor[C]) => Promise<void>;
```
Comment (dckc) [inline packages/portfolio-contract/src/portfolio.flows.ts]: A yield protocol doesn't make much sense without some way to claim rewards, right, does it? Here's hoping I manage to find out more from others.

How about we take out the `?` and have USDN throw for now.

File: packages/portfolio-contract/src/pos-gmp.flows.ts [hunk 7]
```diff
@@ -190,6 +197,14 @@ const Aave: AaveI = {
   withdraw: ['address', 'uint256', 'address'],
 };
 
+type AaveRewardsControllerI = {
+  claimAllRewardsToSelf: ['address[]'];
+};
+
+const AaveRewardsController: AaveRewardsControllerI = {
+  claimAllRewardsToSelf: ['address[]'],
+};
```
Comment (dckc) [inline packages/portfolio-contract/src/pos-gmp.flows.ts:198]: Where did you find those method signatures? Please [cite external constraints](https://github.com/Agoric/agoric-sdk/wiki/Cite-External-Constraints); for example:

https://github.com/Agoric/agoric-sdk/blob/fbca5699394bf6997bb58cd650ddadb19a355239/packages/portfolio-contract/src/pos-gmp.flows.ts#L75-L81

File: packages/portfolio-deploy/src/axelar-configs.js [hunk 8]
```diff
@@ -116,6 +116,50 @@ const usdcAddresses = {
...
+    Avalanche: '0x',
+    Arbitrum: '0x',
+    Optimism: '0x',
+    Polygon: '0x',
+    Fantom: '0x',
+    Binance: '0x',
+  },
+  testnet: {
+    Ethereum: '0x', // Sepolia
+    Avalanche: '0xb1c85310a1b809C70fA6806d27Da425C1261F801', // Fuji
```
Comment (dckc) [inline packages/portfolio-deploy/src/axelar-configs.js:137]: where did this address come from?

**Review by dckc — COMMENTED** [review]
some comments so far

I look forward to tests and some clarity in the PR description

File: packages/portfolio-deploy/src/axelar-configs.js (hunk 8 above)
Comment (frazarshad) [reply packages/portfolio-deploy/src/axelar-configs.js:137]: address is from [this resource ](https://search.onaave.com/?q=fuji)

File: packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md [hunk 9]
```diff
@@ -4,274 +4,7 @@ The actual snapshot is saved in `portfolio.flows.test.ts.snap`.
...
-          totalOut: {
-            brand: Object @Alleged: USDC brand {},
-            value: 0n,
-          },
-        },
-      ],
-    ]
-
-## open portfolio with Aave and USDN positions then inbound GMP
+## open portfolio and claim with Aave position
```
Comment (dckc) [inline packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md]: ugh. the snapshot diffs are huge.

please put new tests at the end

File: packages/portfolio-deploy/src/axelar-configs.js (hunk 8 above)
Comment (dckc) [reply packages/portfolio-deploy/src/axelar-configs.js:137]: I wasn't clear: I was asking for the source of that contract address to go into the code; ideally pointing to some dated decision. But that doesn't seem to be available.

Maybe add https://testnet.snowscan.xyz/ in a comment?

I found it somewhat handy to confirm `aaveUsdcAddresses.Avalanche`:
https://testnet.snowscan.xyz/token/0xb1c85310a1b809C70fA6806d27Da425C1261F801

... though what we got there is a 1st-party assertion (anybody can say "my new token is BTC")

File: packages/portfolio-contract/src/pos-gmp.flows.ts [hunk 10]
```diff
@@ -266,8 +307,23 @@ export const CompoundProtocol = {
     const target = { axelarId, remoteAddress };
     await sendGMPContractCall(target, calls, fee, lca, gmpChain);
   },
+  claimRewards: async (ctx, src) => {
```
Comment (dckc) [inline packages/portfolio-contract/src/pos-gmp.flows.ts]: I wonder if `claimRewards` should be folded into `withdraw` using a `claim: boolean` arg?

The test snapshot shows 2 GMP transfers in a row:

```
     { _method: 'transfer', address: { chainId: 'axelar-6' } },
     { _method: 'transfer', address: { chainId: 'axelar-6' } },
```

The test shows 2 different moves that differ only in `claim: true`.

```ts
        { src: '@Arbitrum', dest: 'Aave_Arbitrum', amount, fee: feeCall },
        { src: '@Arbitrum', dest: 'Aave_Arbitrum', amount, fee: feeCall, claim: true },
```

Can we combine those into 1?

File: packages/portfolio-contract/test/portfolio.flows.test.ts [hunk 11]
```diff
@@ -513,6 +521,53 @@ test('open portfolio with Aave position', async t => {
...
 });
 
+test('open portfolio and claim with Aave position', async t => {
+  const { add } = AmountMath;
+  const amount = AmountMath.make(USDC, 300n);
+  const feeAcct = AmountMath.make(BLD, 50n);
+  const feeCall = AmountMath.make(BLD, 100n);
+  const { orch, tapPK, ctx, offer, storage } = mocks(
+    {},
+    { Deposit: amount, GmpFee: add(feeAcct, feeCall) },
```
Comment (dckc) [inline packages/portfolio-contract/test/portfolio.flows.test.ts]: 2 steps below use `feeCall`. So unless we combine them, I think that should be:


```suggestion
    { Deposit: amount, GmpFee: add(feeAcct, add(feeCall, feeCall)) },
```

File: packages/portfolio-contract/test/portfolio.flows.test.ts [hunk 12]
```diff
@@ -513,6 +521,53 @@ test('open portfolio with Aave position', async t => {
...
+  );
+
+  const [actual] = await Promise.all([
+    openPortfolio(orch, ctx, offer.seat, {
+      flow: [
+        { src: '<Deposit>', dest: '@agoric', amount },
+        { src: '@agoric', dest: '@noble', amount },
+        { src: '@noble', dest: '@Arbitrum', amount, fee: feeAcct },
+        { src: '@Arbitrum', dest: 'Aave_Arbitrum', amount, fee: feeCall },
+        { src: '@Arbitrum', dest: 'Aave_Arbitrum', amount, fee: feeCall, claim: true },
```
Comment (dckc) [inline packages/portfolio-contract/test/portfolio.flows.test.ts]: It seems unlikely that the amount to claim would be the same as the amount to withdraw. I suggest using different amounts.

File: packages/portfolio-contract/test/portfolio.flows.test.ts [hunk 13]
```diff
@@ -513,6 +521,53 @@ test('open portfolio with Aave position', async t => {
...
+      amounts: {
+        Deposit: { value: 300n },
+        GmpFee: { value: 150n },
+      },
+    },
+    { _method: 'transfer', address: { chainId: 'noble-5' } },
+    { _method: 'transfer', address: { chainId: 'axelar-6' } },
+    { _method: 'depositForBurn' },
+    { _method: 'transfer', address: { chainId: 'axelar-6' } },
+    { _method: 'transfer', address: { chainId: 'axelar-6' } },
```
Comment (dckc) [inline packages/portfolio-contract/test/portfolio.flows.test.ts]: I hope we can combine these GMP transfers into 1.

File: packages/portfolio-contract/src/portfolio.flows.ts [hunk 14]
```diff
@@ -270,13 +271,15 @@ type Way =
       poolKey: PoolKey;
       /** chain with account where assets will come from */
       src: SupportedChain;
+      claim?: boolean;
```
Comment (dckc) [inline packages/portfolio-contract/src/portfolio.flows.ts]: This is money going _into_ a pool; i.e. `supply`. `claim` doesn't make sense here, does it?

File: packages/portfolio-contract/src/portfolio.flows.ts [hunk 15]
```diff
@@ -148,6 +148,7 @@ export type ProtocolDetail<
     amount: NatAmount,
     dest: AccountInfoFor[C],
   ) => Promise<void>;
+  claimRewards: (ctx: CTX, src: AccountInfoFor[C]) => Promise<void>;
```
Comment (dckc) [inline packages/portfolio-contract/src/portfolio.flows.ts]: I wonder whether we can avoid this in favor of a `claim` flag on `withdraw`.

**Review by dckc — COMMENTED** [review]
what do you think about folding claim into withdraw?

and please move the new flow test to the end so the snapshot diffs are readable

File: packages/portfolio-contract/src/pos-gmp.flows.ts (hunk 10 above)
Comment (frazarshad) [reply packages/portfolio-contract/src/pos-gmp.flows.ts]: i have one case scenario that might not fit this:
the use _only_ wants to claim rewards i.e. without withdrawing their position

Comment (dckc) [reply packages/portfolio-contract/src/pos-gmp.flows.ts]: `{ src: _pool_, dest: _chain, amount: 0, claim: true }` could say that, I think

Comment (frazarshad) [reply packages/portfolio-contract/src/pos-gmp.flows.ts]: that's a nice way to put it 👍 ill make the update

File: packages/portfolio-contract/test/portfolio.flows.test.ts [hunk 16]
```diff
@@ -710,3 +718,52 @@ test.failing(
...
+  );
+
+  const [actual] = await Promise.all([
+    openPortfolio(orch, ctx, offer.seat, {
+      flow: [
+        { src: '<Deposit>', dest: '@agoric', amount },
+        { src: '@agoric', dest: '@noble', amount },
+        { src: '@noble', dest: '@Arbitrum', amount, fee: feeAcct },
+        { src: '@Arbitrum', dest: 'Aave_Arbitrum', amount, fee: feeCall },
+        { dest: '@Arbitrum', src: 'Aave_Arbitrum', amount: emptyAmount, fee: feeCall, claim: true },
```
Comment (dckc) [inline packages/portfolio-contract/test/portfolio.flows.test.ts]: is it plausible that someone would claim rewards while opening a portfolio?

I suggest putting it in a subsequent `rebalance()`

**Comment by dckc** [comment]
Access Control / Authorization

Who is allowed to call the claim rewards function?
Is there validation to ensure only authorized users/addresses can claim rewards?
Reentrancy

Does the function interact with external contracts or allow callbacks?
Is it protected against reentrancy attacks (e.g., using checks-effects-interactions pattern or reentrancy guards)?
Input Validation
I tried to get CoPilot to help with filling out the PR template. It was mostly just frustration; the thing lost all its marbles eventually; but there are some bits I want to think about further..

## Security Considerations

Are inputs checked for correctness and sanity?
Can malicious input cause unexpected behavior or vulnerabilities?
State Manipulation

Is the contract state updated correctly and atomically?
Could an attacker manipulate the contract state to claim more rewards than allowed?
Race Conditions

Is there any risk that concurrent claims could cause inconsistent state or double-claiming?
Funds Handling

Are rewards transferred securely to the claimant?
Is there protection against sending funds to unintended addresses?
Event Logging

Are important actions (claims, transfers) logged for auditability and monitoring?
Gas/Resource Constraints

Can the function be exploited to consume excessive resources or block contract execution?
Upgradability/Immutability

Will this change affect upgradability or introduce risks if the contract logic is updated in the future?
Interoperability

If interacting with other contracts, are there security assumptions about those contracts?

----

re "Is there validation to ensure only authorized users/addresses can claim rewards?" this question is hardly coherent, given ocap patterns, right?
Copilot said: Yes, you are correct—this question is somewhat
Yes, you are correct—this question is somewhat incoherent in the context of object capability (ocap) patterns, which are foundational in Agoric and similar systems.

Why?

In ocap systems, authority to perform an action (like “claim rewards”) is not determined by checking a user's address against a list of authorized addresses.
Instead, authority is granted by possession of an appropriate object reference (capability). If you have the capability (usually via a reference or invitation), you can perform the action.
There is no global access control list or identity-based permission check; the security model is based on the unforgeability and controlled distribution of capabilities.
So:

The question “Is there validation to ensure only authorized users/addresses can claim rewards?” presumes an identity/ACL model.
In ocap, “validation” is replaced by “can you present the capability/invitation needed for this action?”
If you don’t have the invitation (or capability reference), you simply cannot claim rewards.
Summary:
In an ocap system like Agoric, the question is mostly moot. Capability possession determines authority, not identity/address matching or external validation.

**Comment by dckc** [comment]
## Docs

To trigger a claim, set `claim: true`

**Comment by dckc** [comment]
## Testing

Beyond the unit tests included in the PR, additional testing should focus on the following areas:

Integration Tests:
Ensure that claiming rewards via the rebalance/withdrawal flow works across real contract deployments and interacts correctly with external protocols (e.g., Aave, Compound) in a staging or testnet environment.

Edge Cases:

Test flows where no rewards are available to claim (claim: true but zero rewards).
Test multiple claims in one offer, across different positions/protocols.
Test incorrect or malformed claim requests (e.g., claim: true on unsupported protocols or positions).
Test partial withdrawal with claim: true.
State and History Validation:
Confirm that portfolio status, history, and user-facing records are updated correctly after a claim operation.

CI and Automation Impact:

Ensure that CI reliably executes new and existing tests without flakiness due to asynchronous reward settlement or external RPC dependencies.
Update or add testnet scripts to include reward-claiming flows and observe correct payouts on-chain.
Monitor for regressions: ensure that adding reward claiming does not break unrelated flows (open, rebalance without claim, etc.).
Security and Permissioning:

Test that only authorized portfolio holders can initiate claims.
Verify that claims cannot be triggered maliciously or unintentionally.

## missing docs?

portfolio.flows.test.ts does indeed have many tests with the following pattern:

Prepare mock arguments for openPortfolio
Call openPortfolio to trigger the flow
Snapshot/mock logs: capture the sequence of calls to mocks (e.g., protocol adapters, chain bridges)
Snapshot vstorage output: validate the resulting state in vstorage, typically the portfolio/account record

File: packages/portfolio-contract/src/type-guards-steps.ts [hunk 17]
```diff
@@ -61,6 +61,7 @@ export type MovementDesc = {
   fee?: NatAmount;
   /** for example: { usdnOut: 98n } */
   detail?: Record<string, NatValue>;
+  claim?: boolean;
```
Comment (dckc) [inline packages/portfolio-contract/src/type-guards-steps.ts:68]: The corresponding shape/pattern needs to change.  **important**

See `TypedPattern` discussion in `CONTRIBUTING.md`.

----

Why didn't this get caught by tests?
oh! `openPortfolio` assumes the caller has done pattern validation.

Please add a test that uses `claim: true` in `offer-shapes.test.ts`

**Comment by dckc** [comment]
## Upgrade considerations: none

Upgrade Considerations
This PR introduces a new way to claim protocol rewards by including claim: true in a portfolio flow step (e.g., for Aave positions).
Upgrade impact:

There are no on-chain migrations or changes to persistent contract state.
Existing portfolios and flows remain valid; the feature is additive.
No manual intervention is required for users or operators after upgrade.
The new reward-claiming functionality is only invoked when explicitly requested in a new flow—no existing flows or offers are affected.
Integration and acceptance testing should verify that flows with and without claim: true continue to execute as expected.
Verification Steps:

After deployment, create a test portfolio and execute a withdrawal/rebalance flow with claim: true to confirm rewards are credited.
Confirm that legacy flows (without claim: true) behave as before.
Monitor logs and vstorage for any unexpected errors or side effects.
No special upgrade scripts or coordination needed; standard deployment procedures are sufficient.

You said: that's all true, but is it remarkable?
that's all true, but is it remarkable? does "none" cover it?
Copilot said: No, it’s not remarkable—this is a standard,
No, it’s not remarkable—this is a standard, non-breaking, additive feature with no migration or manual intervention required.
In this case, “none” would adequately answer the Upgrade Considerations prompt.

Example:

Upgrade Considerations:
None.

This is appropriate for features that do not impact upgrade complexity, existing state, or require operator/user action.

File: packages/portfolio-contract/test/portfolio.flows.test.ts [hunk 18]
```diff
@@ -710,3 +718,52 @@ test.failing(
     }),
   ],
 );
+
+test('open portfolio and claim with Aave position', async t => {
```
Comment (dckc) [inline packages/portfolio-contract/test/portfolio.flows.test.ts]: please add a test for what happens if you put `claim: true` on a withdraw from a USDN position.

I expect it to throw.

File: packages/portfolio-contract/src/type-guards-steps.ts (hunk 17 above)
Comment (dckc) [reply packages/portfolio-contract/src/type-guards-steps.ts:68]: Rather, let's be sure to do a contract test not only to catch this problem with the shape but to validate the whole user story; i.e. add a test to `portfolio.contract.test.ts`.

tests in `offer-shapes.test.ts` are then nice-to-have.

File: packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md [hunk 19]
```diff
@@ -1231,3 +1231,170 @@ Generated by [AVA](https://avajs.dev).
...
...0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,202,199,255,168,44,15,67,235,176,252,17,252,211,33,35,236,164,102,38,207,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,64,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,68,9,94,167,179,0,0,0,0,0,0,0,0,0,0,0,0,135,135,11,202,63,63,214,51,92,63,76,232,57,45,105,53,11,79,164,226,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,44,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,135,135,11,202,63,63,214,51,92,63,76,232,57,45,105,53,11,79,164,226,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,64,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,132,97,123,160,55,0,0,0,0,0,0,0,0,0,0,0,0,202,199,255,168,44,15,67,235,176,252,17,252,211,33,35,236,164,102,38,207,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,44,0,0,0,0,0,0,0,0,0,0,0,0,18,108,243,172,158,161,39,148,255,80,245,103,39,199,198,110,38,217,192,146,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"type":1,"fee":{"amount":"100","recipient":"axelar1zl3rxpp70lmte2xr6c4lgske2fyuj3hupcsvcd"}}',
```
Comment (dckc) [inline packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md]: the payload isn't really readable here.

What do you think about an additional test just for the payload?
Or for the `calls`?

File: packages/portfolio-deploy/src/axelar-configs.js [hunk 20]
```diff
@@ -116,6 +116,50 @@ const usdcAddresses = {
   },
 };
 
+/** @type {AddressesMap} */
+const aaveUsdcAddresses = {
+  mainnet: {
+    Ethereum: '0x',
```
Comment (dckc) [inline packages/portfolio-deploy/src/axelar-configs.js:127]: please add a TODO about filling in the rest of these addresses. include an issue number. make a new issue if necessary

**important**

File: packages/portfolio-contract/src/portfolio.flows.ts [hunk 21]
```diff
@@ -296,7 +298,12 @@ export const wayFromSrcToDesc = (moveDesc: MovementDesc): Way => {
         !feeRequired.includes(protocol) ||
         Fail`missing fee ${q(moveDesc)}`;
       // XXX check that destName is in protocol.chains
-      return { how: protocol, poolKey, dest: destName };
+      return {
+        how: protocol,
+        poolKey,
+        dest: destName,
+        claim: Boolean(moveDesc.claim),
```
Comment (dckc) [inline packages/portfolio-contract/src/portfolio.flows.ts]: why `Boolean(...)`? The caller ensures the data is pattern-validated. As noted elsewhere, the relevant pattern/shape does need updating. **important**

```suggestion
        claim: moveDesc.claim,
```

File: packages/portfolio-contract/test/portfolio.flows.test.ts [hunk 22]
```diff
@@ -75,12 +75,20 @@ const destinationEVMChain: AxelarChain = 'Arbitrum';
...
+const toHandler = new Map();
 const mockZCF: ZCF = Far('MockZCF', {
   makeEmptySeatKit: () =>
+    // @ts-expect-error mock
     ({
       zcfSeat: Far('MockZCFSeat', { exit: theExit }),
     }) as unknown as ZCF,
+
+  // @ts-expect-error mock
+  makeInvitation: async (handler, _desc, _c, _patt) => {
```
Comment (dckc) [inline packages/portfolio-contract/test/portfolio.flows.test.ts]: I don't see where this is used; why is it added?
**important**

File: packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md [hunk 23]
```diff
@@ -1231,3 +1231,170 @@ Generated by [AVA](https://avajs.dev).
...
+    [
+      [
+        'published.ymax0.portfolios.portfolio1',
+        {
+          accountIdByChain: {
+            Arbitrum: 'eip155:42161:0x126cf3AC9ea12794Ff50f56727C7C66E26D9C092',
+            agoric: 'cosmos:agoric-6:agoric11014',
+            noble: 'cosmos:noble-5:noble11028',
+          },
+          flowCount: 1,
```
Comment (dckc) [inline packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md:1402]: let's have 2 flows: 1 to open, 1 to claim+withdraw

**Review by dckc — CHANGES_REQUESTED** [review]
closer all the time...
just a few **important** things remain

File: packages/portfolio-deploy/src/axelar-configs.js (hunk 20 above)
Comment (frazarshad) [reply packages/portfolio-deploy/src/axelar-configs.js:127]: created this issue https://github.com/Agoric/agoric-sdk/issues/11651

File: packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md [hunk 24]
```diff
@@ -1231,3 +1231,126 @@ Generated by [AVA](https://avajs.dev).
...
+        },
+      },
+      {
+        _cap: 'seat',
+        _method: 'exit',
+        completion: undefined,
+      },
+    ]
+
+> decoded calls
```
Comment (dckc) [inline packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md:1368]: handy; thanks!

File: packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md [hunk 25]
```diff
@@ -1231,3 +1231,126 @@ Generated by [AVA](https://avajs.dev).
         },
       ],
     ]
+
+## open portfolio and claim with Aave position
```
Comment (dckc) [inline packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md]: "open portfolio" is no longer part of this test, right?

File: packages/portfolio-contract/test/utils.ts [hunk 26]
```diff
@@ -0,0 +1,47 @@
+import { decodeAbiParameters, decodeFunctionData } from 'viem';
```
Comment (dckc) [inline packages/portfolio-contract/test/utils.ts]: The name `utils.ts` is kinda unhelpful. Please consider `gmp-utils.ts` or `evm-call-utils.ts`.

File: packages/portfolio-deploy/src/axelar-configs.js (hunk 20 above)
Comment (dckc) [reply packages/portfolio-deploy/src/axelar-configs.js:127]: It's important that we put it on the project board. I just added it, this time.

File: packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md (hunk 25 above)
Comment (frazarshad) [reply packages/portfolio-contract/test/snapshots/portfolio.flows.test.ts.md]: updated

File: packages/portfolio-contract/test/snapshots/portfolio.contract.test.ts.md [hunk 27]
```diff
@@ -379,3 +379,81 @@ Generated by [AVA](https://avajs.dev).
         value: 0n,
       },
     }
+
+## USDN claim fails currently
```
Comment (dckc) [inline packages/portfolio-contract/test/snapshots/portfolio.contract.test.ts.md]: no need to snapshot this one.

File: packages/portfolio-contract/test/portfolio.contract.test.ts [hunk 28]
```diff
@@ -239,3 +239,129 @@ test('open portfolio with USDN, Aave positions', async t => {
   t.snapshot(contents, 'vstorage');
   t.snapshot(done.payouts, 'refund payouts');
 });
+
+test('claim rewards on Aave position successfully', async t => {
```
Comment (dckc) [inline packages/portfolio-contract/test/portfolio.contract.test.ts:294]: snapshot this one. **important**

**Review by dckc — COMMENTED** [review]
for the contract test, snapshot the working one, not the unsupported one

File: packages/portfolio-contract/src/portfolio.flows.ts [hunk 29]
```diff
@@ -366,7 +373,7 @@ const stepFlow = async (
     const gInfo = await provideEVMAccount(chain, gmp, lca, ctx, kit);
     const accountId: AccountId = `${gInfo.chainId}:${gInfo.remoteAddress}`;
 
-    const evmCtx: EVMContext<N> = harden({
+    const evmCtx = harden({
```
Comment (dckc) [inline packages/portfolio-contract/src/portfolio.flows.ts]: why take out the type annotation?

File: packages/portfolio-contract/test/portfolio.flows.test.ts [hunk 30]
```diff
@@ -55,6 +55,7 @@ import {
   makeIncomingEVMEvent,
   makeIncomingVTransferEvent,
 } from './supports.ts';
+import { decodeFunctionCall } from './abi-utils.ts';
```
Comment (dckc) [inline packages/portfolio-contract/test/portfolio.flows.test.ts:63]: thanks; `abi-utils` is much more helpful

**Review by dckc — APPROVED** [review]
well done!
